annealing_helper_functions.py

```diff
@@ -76,3 +77,55 @@ def simulated_annealing(graph, function, initial_X, initial_temp, cool, reanneal
                 T = 1.
             
     return X, history
+
+def prob(a, b):
+    return np.exp(a / b)
+
+# Adapted from Lecture14_Parallel_Tempering_and_Emcee.ipynb
+def parallel_tempering(graph, function, initial_Xs, initial_temps, 
+                       iterr, swap_function, nswaps, nbefore):
+    # Make sure inputs are ok
+    assert(len(initial_temps) == len(initial_Xs)), "Mismatched input dimensions"
+    assert(initial_temps[0] == 1), "First temperature should be one"
```

Comment:
I did this because we need to return the system with temperature one, and given lots of systems, it's a slight slight slight performance boost to not have to search through all the temperatures and just return the first one. In retrospect, it would definitely be better to just search the temperatures or pass in the index of the thread to return.